prov/gni/src/gnix_atomic.c

```diff
@@ -662,6 +662,9 @@ ssize_t _gnix_atomic(struct gnix_fid_ep *ep,
 
 	COND_RELEASE(ep->requires_lock, &ep->vc_lock);
 
+	/* If a new VC was allocated, progress CM before returning. */
```

Comment:
A VC is looked up for the operation.  If it doesn't exist, it's created and connection setup is initiated.

vc_progress's only role right now is to conditionally call cm_nic_progress.  cm_nic_progress cannot be called with the VC's lock held.  You stir up a good point, actually, since I shouldn't actually be testing vc->conn_state without the VC's lock held.  Maybe I should test the VC's conn state before releasing the lock, then call progress if needed after dropping the lock?